### ISSUE TITLE
Bug: confusing errors when using flat config without enabling it

### ISSUE THREAD

**Issue by kkmuffme** [eslint/eslint#17369]
### Environment

Node version: 16.19.0
npm version: 8.9.0
Local ESLint version: 8.45.0
Global ESLint version: 8.45.0
Operating System: RHEL

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
module.exports = [
    {
        rules: {
            semi: "error",
            "prefer-const": "error"
        }
    }
];
```
</details>

```js
var foo = 'hello'
```

### What did you expect to happen?

When using a flat config file without having set e.g. `ESLINT_USE_FLAT_CONFIG=true` the errors are extremely cryptic and without extensive knowledge of eslint, it's hard (impossible?) to figure out what the actual issue is (= using a flat config file without having flat config enabled)

### What actually happened?

e.g. for `eslint --format json --no-config-lookup --config some/eslint.config.js`

>Invalid option '--config-lookup' - perhaps you meant '--config'?

>Error: ESLint configuration in --config is invalid:
        - Property "" is the wrong type (expected object but got `[{"rules":{"semi":"error","prefer-const":"error"}}]`).

### Link to Minimal Reproducible Example

see above, example from https://eslint.org/docs/latest/use/configure/configuration-files-new

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Rec0iL99** [comment]
Hi @kkmuffme, thanks for the issue. I think the 

```bash
Error: ESLint configuration in --config is invalid:
- Property "" is the wrong type (expected object but got [{"rules":{"semi":"error","prefer-const":"error"}}]).
```

part is happening because you haven't set the environment variable `ESLINT_USE_FLAT_CONFIG `. ESLint expects you to set the environment variable `ESLINT_USE_FLAT_CONFIG ` to `true`, if you don't have `eslint.config.js` in the root of the project.

1. I put the `eslint.config.js` file in a folder `some/`

```js
// some/eslint.config.js
module.exports = [
    {
        rules: {
            semi: "error",
            "prefer-const": "error"
        },
        foo: { }
    }
];
``` 

**Result:**

```bash
npx eslint --config some/eslint.config.js index.js                                                                                                                                                                 

Oops! Something went wrong! :(

ESLint: 8.45.0

Error: ESLint configuration in --config is invalid:
        - Property "" is the wrong type (expected object but got `[{"rules":{"semi":"error","prefer-const":"error"},"foo":{}}]`).

    at ConfigValidator.validateConfigSchema (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2168:19)
    at ConfigArrayFactory._normalizeConfigData (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3010:19)
    at ConfigArrayFactory._loadConfigData (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2975:21)
    at ConfigArrayFactory.loadFile (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2841:40)
    at createCLIConfigArray (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3649:35)
    at new CascadingConfigArrayFactory (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3724:29)
    at new CLIEngine (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/eslint/lib/cli-engine/cli-engine.js:617:36)
    at new ESLint (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/eslint/lib/eslint/eslint.js:430:27)
    at Object.execute (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/eslint/lib/cli.js:382:24)
    at async main (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/eslint/bin/eslint.js:135:24)
```

**Result with `ESLINT_USE_FLAT_CONFIG=true`:**

```bash
ESLINT_USE_FLAT_CONFIG=true npx eslint --config some/eslint.config.js index.js                                                                                                                                     

Oops! Something went wrong! :(

ESLint: 8.45.0

Error: Unexpected key "foo" found.
    at ObjectSchema.validate (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/object-schema/src/object-schema.js:203:23)
    at /Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/object-schema/src/object-schema.js:171:18
    at Array.reduce (<anonymous>)
    at ObjectSchema.merge (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/object-schema/src/object-schema.js:169:24)
    at /Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/config-array/api.js:916:42
    at Array.reduce (<anonymous>)
    at FlatConfigArray.getConfig (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/config-array/api.js:915:39)
    at FlatConfigArray.isFileIgnored (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/config-array/api.js:943:15)
    at /Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/eslint/lib/eslint/eslint-helpers.js:504:38
    at Array.forEach (<anonymous>)
```

2. I moved the `some/eslint.config.js` to the `root` of the project.

**Result:**

```bash
npx eslint --config eslint.config.js index.js 
                                                                                                                                                                     
Oops! Something went wrong! :(

ESLint: 8.45.0

Error: Unexpected key "foo" found.
    at ObjectSchema.validate (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/object-schema/src/object-schema.js:203:23)
    at /Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/object-schema/src/object-schema.js:171:18
    at Array.reduce (<anonymous>)
    at ObjectSchema.merge (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/object-schema/src/object-schema.js:169:24)
    at /Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/config-array/api.js:916:42
    at Array.reduce (<anonymous>)
    at FlatConfigArray.getConfig (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/config-array/api.js:915:39)
    at FlatConfigArray.isFileIgnored (/Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/@humanwhocodes/config-array/api.js:943:15)
    at /Users/joelmathew/WebProjects/eslint-misc/eslint-flat-no-set/node_modules/eslint/lib/eslint/eslint-helpers.js:504:38
    at Array.forEach (<anonymous>)
```

For now, ESLint strictly needs to have the `ESLINT_USE_FLAT_CONFIG` variable set for expected results to happen if `eslint.config.js` is not at the root of the project. This will change when we fully migrate to flat config in `v9`.

**Comment by Rec0iL99** [comment]
I agree. Sorry, that I missed pushing the issue forward. Thanks

**Comment by nzakas** [comment]
As I mentioned in your other issue, we really can't tell if you're using a flat config in eslintrc mode, all we can do is look for keys that don't match and warn on them. We are trying to catch areas where there might be some confusion, so I'll leave this open in case anyone wants to take a look.

**Comment by nzakas** [comment]
We actually did add in a bunch of validation and clearer errors messages when we are running in flat config mode and get an eslintrc config, but we don't have the opposite. We probably won't get back around to doing that because flat config will be the default in v9.
